torchtune/utils/_distributed.py

```diff
@@ -242,6 +250,83 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
     return lora_wrap_fsdp
 
 
+def load_from_full_model_state_dict(
+    model: FSDPModule,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+    is_rank_zero: bool,
+):
+    meta_sharded_sd = model.state_dict()
+    sharded_sd = {}
+    for param_name, full_tensor in full_sd.items():
+        sharded_meta_param = meta_sharded_sd.get(param_name)
+        # `.to(dtype)` ensures same dtype when `assign=True`
+        sharded_tensor = distribute_tensor(
+            full_tensor.to(sharded_meta_param.dtype),
+            sharded_meta_param.device_mesh,
+            sharded_meta_param.placements,
+        )
+        sharded_sd[param_name] = nn.Parameter(sharded_tensor)
+    # choose `assign=True` since we cannot call `copy_` on meta tensor
+    model.load_state_dict(sharded_sd, strict=False, assign=True)
```

Comment:
If we catch missing and unexpected keys from `load_state_dict` with `strict=False` what format will the keys be in? Previously with FSDP1 the keys contained all the info about FSDP wrapping. E.g.`model.layers.0._fsdp_flat_param.attn.q_proj.weight` (probably not exactly right but something like that). Will that still be the case here?

Reply:
for FSDP2, it's clean FQNs without FSDP prefix. For example, `layers.0.attn.q_proj.lora_a.weight`

FSDP2 is clean because 1) `fully_shard` register hooks instead wrap `nn.Module`,  2) `fully_shard` set `module.__class__ =  type(f"FSDP{cls.__name__}", (FSDPModule, cls), dct)`  https://fburl.com/i20yr3s2

Reply:
This is great! I think this means we can actually do validation of LoRA state dict load more cleanly (note that we actually have two separate utilities for this for the [single-device](https://github.com/pytorch/torchtune/blob/0c4056a6a55bfffd3cec7a99a13dc978cfe85963/recipes/lora_finetune_single_device.py#L282) vs [distributed](https://github.com/pytorch/torchtune/blob/0c4056a6a55bfffd3cec7a99a13dc978cfe85963/recipes/lora_finetune_distributed.py#L304) case because of this FSDP prefix issue). Not a concern for this PR but this will allow us to clean up our code a bit